File: Expt 2/d.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h>
#include<string.h>
#define MAX 100

long int lstack[MAX];
float fstack[MAX];
char prefix[MAX];
int top = -1;
int floatint;

void push(float);
float pop();
void evalprefix();
int isfull();
int isempty();

int isempty(){
	if(top == -1)
		return 1;
	else return 0;
}

int isfull(){
	if(top == (MAX - 1))
		return 1;
	else return 0;
}

void push(float symbol){
	if(isfull()){
		printf("Stack overflow\n");
		return;
	}
	if (floatint == 1)
		fstack[++top] = symbol;
	else lstack[++top] = (long int)symbol;
}

float pop(){
	if(isempty()){
		printf("Stack underflow\n");
		exit(1);
	}
	if (floatint == 1)
		return(fstack[top--]);
	else return((long int)lstack[top--]);
}

void evalprefix(){
	float a, b;
	float temp;
	for(int i = strlen(prefix); i > 0; i++){
		if(prefix[i] >= '0' || prefix[i] <= '9')
			push(prefix[i] - '0');
		else{
			a = pop();
			if (floatint == 1) {
				printf("%f\t", a);
   				for (int j = 0; j <= top; j++)
        		printf("%f,", fstack[j]);
			} else {
				printf("%ld\t", a);
    			for (int j = 0; j <= top; j++)
        			printf("%ld,", lstack[j]);
			}
			printf("\n");
			b = pop();
			if (floatint == 1) {
				printf("%f\t", b);
   				for (int j = 0; j <= top; j++)
        		printf("%f,", fstack[j]);
			} else {
				printf("%ld\t", b);
    			for (int j = 0; j <= top; j++)
        			printf("%ld,", lstack[j]);
			}
			printf("\n");
			switch(prefix[i]){
				case '+': temp = a + b;
					break;
				case '-': temp = a - b;
					break;
				case '*': temp = a * b;
					break;
				case '/': temp = a / b;
					break;
				case '%': temp = (long int)a % (long int)b;
					break;
				case '^': temp = 1;
					for(int i = 1; i <= a; i++)
						temp = temp * b;
			}
			push(temp);
		}
	}
}

int main(){
	printf("Enter the prefix expression\n");
	scanf("%s", prefix);
	do{
		printf("\nDo you want it as a 1. Float or 2. Integer?: ");
		scanf("%d", &floatint);
		if (floatint != 1 && floatint != 2)
			printf("Invalid input\n");
	} while (floatint != 1 && floatint != 2);
	evalprefix();
	if(floatint == 1)
		printf("The evaluation is: %f\n", fstack[0]);
	else 
		printf("The evaluation is: %f\n", lstack[0]);
	return 0;
}

File: Expt 2/d.py
def is_operand(c):
    return c.isalpha()

def evaluate_operator(op, operand1, operand2):
    if op == '+':
        return operand1 + operand2
    elif op == '-':
        return operand1 - operand2
    elif op == '*':
        return operand1 * operand2
    elif op == '/':
        return operand1 / operand2
    elif op == '^':
        return operand1 ** operand2

def evaluate_prefix(expression, variable_values):
    stack = []
    expression = expression[::-1]  # Reverse the expression to start from the beginning

    for symbol in expression:
        if is_operand(symbol):
            stack.append(variable_values[symbol])
        else:
            operand1 = stack.pop()
            operand2 = stack.pop()
            result = evaluate_operator(symbol, operand1, operand2)
            stack.append(result)
        
        # Display the current step
        stack_display = ['%.6f' % item if isinstance(item, float) else item for item in stack]
        print(f"{symbol}\t{' '.join(map(str, stack_display))}")

    return stack[0]  # The final result is at the top of the stack

# Variable values
variables = {
    'A': 3,
    'B': 2,
    'C': 2,
    'D': 2,
    'E': 2,
    'F': 1
}

expression = "+*A-AE*B^^FB-+ECE"
print("Enter the prefix expression: ", expression)
print("\nDo you want the result as a: 1. Float or 2. Integer?: 2")
print("Symbol\tStack")
result = evaluate_prefix(expression, variables)
print("The evaluation is", result)
